### ISSUE TITLE
API concerns:  Atomic operations, putIfAbsent, void returning put / delete

### ISSUE THREAD

**Issue by scottcarey** [lmdbjava/lmdbjava#31]
lmdb looks very interesting, but the feature set exposed in APIs like this one is woefully lacking if I were to try and replace my existing solution with it.

The C API supports a put with the MDB_NOOVERWRITE flag, but this API throws an exception (awkward) if there is a value.  I would expect that users of this flag actually want either a boolean or the existing value returned instead of an error.  One could catch the exception, then issue a get() to get the value, but that is probably significantly slower --- the exception handling on one side, and the additional get traversing the index on the other, when it could all be done in a single index traversal.

Ideally one would have a `putIfAbsent(key, value)` method.   This needs to return (a pointer to) the prior value, if present.

Ideally ordinary put supports easily retrieving the value that was replaced as well.   From what I can tell, it should be possible to move the cursor to a key (if it exists), fetch the old value, and put the new one in, atomically, for both 'put' and 'putIfAbsent' use cases.  LMDBs concurrency model seems strong enough to do so.

Additionally supporting conditional replace -- where the put succeeds atomically only if the prior value matches some expectation, would be of great benefit for some use cases.

I have an application where multiple concurrent threads are updating data, and I need to atomically capture every state transition.  This is fairly easy with any store that supports atomic operations and virtually impossible otherwise.

Basically, I want ConcurrentMap: http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentMap.html

And I want delete / put to not return void, since there are valid non-exceptional results for both of those.   In the case of delete -- did it actually delete something?   In the case of put -- was there a prior value?

**Comment by benalexau** [comment]
I agree we should modify `Dbi.put(..)` so if invoked with `MDB_NOOVERWRITE` and LMDB returns result code `MDB_KEYEXIST` we should set the `Txn.val()` to point to the existing item. This aligns with the guarantees offered by the underlying `mdb_put`. I note there is no such guarantee offered by the equivalent cursor method (`mdb_cursor_put`).

Your issue also requested support for returning the prior value in successful mutation cases. While seemingly straight-forward, at the moment LmdbJava does not perform any buffer copying. Specifically, put operations rely on the user to allocate an appropriately-sized buffer for their operation. For get operations we modify the relevant buffer instance to point at the LMDB-managed memory address and size.

While we could add a buffer copying mechanism to our `BufferProxy` API, this would mean supporting it in each implementation (ie Netty, Agrona, Java `ByteBuffer`). Furthermore, we wouldn't be able to easily control the lifecycle of the copied buffer without coupling it to `Txn` (eg by adding additional fields for "secondary value" or similar) or simply giving up on using an allocation/deallocation pool for these copied values. Indeed given we need to allocate memory for the copy, it would make more sense to allocate a new buffer for the required prior value copy. However this approach would be slow due to the object allocation and GC overhead.

As such it's not as straight-forward as it may seem to return the existing value when overwriting. Users needing this can use a `Cursor` to efficiently perform a get within a single `Txn` boundary, focus on the existing value, do whatever comparison they wish, copy the existing value to another buffer (eg one they control the broader lifecycle of and can select a maximum size for their data that permits a single allocation and buffer reuse) and then mutate the value if needed.

I'm open to any higher-level interface that makes it easier to use LmdbJava like a traditional concurrent `Map`, but I think we should have it as a separate type that imposes some documented limitations that would be necessary (eg performance due to existing value copying, maximum value size for buffer pre-allocation, copy buffer reference lifecycle constraints, existing or new transaction participation, compatible buffer type(s), compatible DBI flags etc).

**Comment by benalexau** [comment]
Did the OP have any further comments on this ticket? 

Basically a higher-level API is being requested that exposes CAS and existing value copy semantics. This is certainly possible (as the `Txn` provides isolation) but an implementation must make trade-offs for each of the limitations mentioned in my last comment. One project's acceptable trade-offs will differ from another, so it is difficult to come up with a broadly-acceptable higher-level API.

For example, say we decided to simplify things a little by `DurableMap`:

* Only supporting `ByteBuffer`
* Requiring all methods to present an non-closed `Txn`
* Being constructed with a specific `Dbi`

So far so good, but let's consider the `replace` use case and how do we return the existing value being overwritten. Where do we get a buffer to hold that value? Some obvious alternatives are:

1. `void replace(Txn txn, ByteBuffer key, ByteBuffer value)`: In this case the existing value isn't ever returned, making it allocation-free, copy-free and fast. But we already have this with `Dbi` and `Cursor` methods right now.
2. `ByteBuffer replaceAndGetExisting(Txn txn, ByteBuffer key, ByteBuffer value)`: Here a new `ByteBuffer` must be allocated and the existing value copied into it. This will be slow.
3. `void replaceAndGetExisting(Txn txn, ByteBuffer key, ByteBuffer value, ByteBuffer existingHolder)`: In this case the user presents a buffer to hold the existing value. The copy cost is still incurred, but the signature has become overwhelming. Plus, what should happen if the existing value is too large to fit in the`existingHolder` (abort, truncate)?

All of this can figured out, but it's not easy to develop a clean, high-level API that retains high performance and suits every project. The existing `Cursor` API already offers the most efficient way to access LMDB-stored data and allows each project to make their own trade-offs.

I'd propose we close this ticket unless someone wants to discuss a higher-level API in detail.

**Comment by scottcarey** [comment]
Yes, I have several comments but have been very busy.

Here is the most complicated part of the API I am trying to implement (that currently has CHM, TrieMap, and MapDB implementations):

````scala
  /**
   * Performs an atomic insert or update.  If the key exists, uses the update function as in
   * `atomicUpdate`.  If the key does not exist, adds the value if absent, as in `putIfAbsent`.
   *
   * The equivalent of
   *
   *     if (containsKey(key)) {
   *       atomicUpdate(key, update)
   *     } else {
   *       putIfAbsent(key, insert())
   *     }
   *
   * Except the entire operation is atomic and the index is traversed only once.
   * Due to concurrent updates, insert and/or update may be called multiple times.
   *
   * Returns `true` if the operation resulted in an update, false otherwise.
   */
  def atomicInsertOrUpdate(key: K, insert: => V, update: V => Option[V]): Boolean

   /**
   * Performs an atomic insert or update.  If the key exists, uses the update function as in
   * `atomicUpdateVal`.  If the key does not exist, adds the value if absent, as in `putIfAbsentVal`.
   *
   * The equivalent of
   *
   *     if (containsKey(key)) {
   *       atomicUpdateVal(key, update)
   *     } else {
   *       (null, putIfAbsentVal(key, insert()))
   *     }
   *
   * Except the entire operation is atomic and the index is traversed only once.
   * Due to concurrent updates, insert and/or update may be called multiple times.
   *
   * Returns `(priorValue, Some(updatedValue))` if the operation resulted in an update
   *   an `updatedValue` of null means the key was removed
   * Returns `(null, Some(insertedValue))` if the operation resulted in an insert of a new value
   * Returns `(priorValue, None)` if the operation did not result in an insert or update.
   */
  def atomicInsertOrUpdateVal(key: K, insert: => V, update: V => Option[V]): (V, Option[V])
````

Note that the above methods are only available from within an enclosing transaction scope.  The former is an optimization of the latter, for cases where serializing all the results is wasteful and unnecessary.

Once serialization enters the picture, method signatures passing a bunch of `ByteBuffer`s becomes wasteful, as you not only want to avoid allocation, you want to avoid serialization when unnecessary.  Why copy a buffer only to be read into a value and discard it when you could have read it into a value prior to the copy?
What is important to provide is what to do with the bytes (if available) or how to produce them (if needed) -- functions to run.

I could use Cursor for everything, maybe.  I think that most use cases require allocation in the case that the user wants data back, and avoiding it is a bit silly.  Yes, there are use cases where you can peek at the raw data without allocation, and the current API supports those use cases already.  A mid-level api (or 'advanced high level') could also support this -- serialization is merely the application of the bytes to producing a value on the heap, other things could be done instead.   But if you want to turn the value into a `Foo`, you are going to allocate.  Either directly by allocating a `Foo` or by wrapping a buffer that can exist outside of transaction scope.  There are many tricks in JVM-land to make this more efficient and low impact on the GC.

When I think about this project vs MapDB, what really is missing is a higher level API (and vice versa for MapDB -- low level API support is poor forcing some inefficiency and unnecessary data copies). 

I may have some time before the year is over to code up a proposal.  My main problem is that I don't understand the low level details of LMDB enough, nor the buffer management in this project, to know that any implementation is correct.

I have two higher level APIs that abstract over various data stores.   The above code is from a atomic/transactional read-write API specialized to one project,  and I also have one optimized for read-only use cases (actually write once, close/optimize the store, then re-load the dataset read-only elsewhere).

Some comments re: `replaceAndGetExisting`

#2 is OK, allocating and copying a value into it is not slow, for this use case.  There is no alternative, if you want the value back.  

#3 is asking the wrong thing of the user, as evidenced by wondering what to do if the value does not fit in the user's buffer. 

What could be faster?  possibly something like

````java
<A> A replaceAndGetExisting(Txn txn, ByteBuffer key, ByteBuffer value, DataReader<A> onReplace);
````

`DataReader` is some sort of function type that represents what to do with the replaced value if and only if it exists.  This can be generic so that the Argona, Netty, and JVM buffer flavors could be supported.  Basically, this is just a code inversion, allowing the user to peek at the data that was replaced inside the cursor and decide what `A` to return as a result -- a minimized set of what is available in the Cursor API bound to the scope of reading one value.  This could be an allocation, a ByteBuffer copy, or it could be a Boolean.  A higher level serialization API would wrap this API and convert K -> ByteBuffer, V -> ByteBuffer, and handle reading the replaced value into a V inside the DataReader supplied.

**Comment by benalexau** [comment]
The ideal API depends on the user's preferred serialization approach. Most Java serialization libraries will go and create a new object to represent the data. So there's an allocation cost there, and a copy cost to get the data into the new object's fields. In such situations I agree there's downside to also creating a temporary `ByteBuffer` and copying data into it. Plus such serializers tend to deal with variable-length objects.

I use Simple Binary Encoding (SBE), which is somewhat different than most Java serializers. Flyweights are used and fixed-length buffers are the norm. I even have a Java annotation processor that generates "framing" objects that hold the optimal-length buffer for each flyweight and provide efficient buffer copy operations for use outside of transactions (plus other goodies like visitors too). So my variant 3 would be useful for a serializer like SBE which typically uses a fixed-length buffer for each message header and message.

On CAS semantics, maybe it's worth mentioning at this point that LMDB only supports one write transaction at a time. A write transaction is free to read data and write data. Other write transaction threads will simply block until the write transaction is available. 

The simplest way to achieve your goals is perhaps to rely on LMDB blocking concurrent writes. For example, a `DurableMap` that is constructed with a `Dbi` instance that the user is responsible for  providing. Individual methods can then use a `ThreadLocal` to fetch a write `Txn`. Use a `Dbi` to CAS the data you need, copying as optimal for your serialization library. When the method that put the `Txn` into the `ThreadLocal` concludes, have it commit or abort the write `Txn`. I think this would work and competing threads would just block when awaiting a new LMDB write transaction.

**Comment by benalexau** [comment]
@scottcarey did you end up solving your use case? If so can we please close this ticket?

**Comment by scottcarey** [comment]
@benalexau I did solve my use case, by writing a higher level API for my app that uses the cursor API to get what it needs done. That only worked after a couple other recent changes.

I'll close this issue.  I may have future contributions as a result of my experience here.

FWIW, It ended up slower than MapDB, because it lacks any caching layer of decoded objects on the heap leading to more deserialization overhead, the single writer lock, and higher I/O per second and wait time.  I am not sure what my ideal configuration is in relation to durability -- I do not need durability on completing a transaction if the app dies or is kill-9'd, e.g. I need ACI, but not D; and if the OS dies or kernel panics, I'm ok with a corrupted database.  The data can be re-generated but it is expensive, and operationally OS panics / crashes are rare and it is acceptable to re-initialize in that case, but app crashes and/or kill-9 unsafe shutdown is more common.  The right durability configuration may reduce my I/O wait time and fix the performance situation.
